apps/emqx_resource/src/emqx_resource_cache.erl

```diff
@@ -0,0 +1,331 @@
+    error :: term(),
+    status :: channel_status(),
+    extra = []
+}).
+
+-define(CB_PT_KEY(ID), {?MODULE, ID}).
+
+new() ->
+    emqx_utils_ets:new(?CACHE, [
+        ordered_set,
+        public,
+        {read_concurrency, true},
+        {keypos, 2}
+    ]).
+
+-spec write(pid(), binary(), resource_data()) -> ok.
+write(ManagerPid, Group, Data) ->
+    #{
+        id := ID,
+        mod := Mod,
+        callback_mode := CallbackMode,
+        query_mode := QueryMode,
+        config := Config,
+        error := Error,
+        state := State,
+        status := Status,
+        added_channels := AddedChannels
+    } = Data,
+    Cb = #{
+        mod => Mod,
+        callback_mode => CallbackMode,
+        query_mode => QueryMode,
+        state => State
+    },
+    IsDryrun = emqx_resource:is_dry_run(ID),
+    Connector = #connector{
+        id = ID,
+        group = Group,
+        manager_pid = ManagerPid,
+        st_err = #{
+            status => Status,
+            error => external_error(Error)
+        },
+        config = Config,
+        cb =
+            case IsDryrun of
+                true ->
+                    %% save callback state in ets for dryrun
+                    Cb;
+                false ->
+                    undefined
+            end,
+        extra = []
+    },
+    Channels = lists:map(fun to_channel_record/1, maps:to_list(AddedChannels)),
+    %% erase old channels (if any)
+    ok = erase_old_channels(ID, maps:keys(AddedChannels)),
+    %% put callback state in persistent_term
+    case IsDryrun of
+        true ->
+            %% do not write persistent_term for dryrun
+            ok;
+        false ->
+            ok = put_state_pt(ID, Cb)
+    end,
+    %% insert connector and channel states
+    true = ets:insert(?CACHE, [Connector | Channels]),
+    ok.
+
+%% @doc Read cached pieces and return a externalized map.
+%% NOTE: Do not call this in hot-path.
+%% TODO: move `group' into `resource_data()'.
+-spec read(resource_id()) -> [{resource_group(), resource_data()}].
+read(ID) ->
+    case ets:lookup(?CACHE, ID) of
+        [] ->
+            [];
+        [#connector{group = G} = C] ->
+            Channels = find_channels(ID),
+            [{G, make_resource_data(ID, C, Channels)}]
+    end.
+
+-spec read_status(resource_id()) -> not_found | st_err().
+read_status(ID) ->
+    ets:lookup_element(?CACHE, ID, #connector.st_err, not_found).
+
+-spec read_manager_pid(resource_id()) -> not_found | pid().
+read_manager_pid(ID) ->
+    ets:lookup_element(?CACHE, ID, #connector.manager_pid, not_found).
+
+-spec read_mod(resource_id()) -> not_found | {ok, module()}.
+read_mod(ID) ->
+    case get_cb(ID) of
+        undefined ->
+            not_found;
+        #{mod := Mod} ->
+            {ok, Mod}
+    end.
+
+get_cb(ID) ->
+    case get_cb_pt(ID) of
+        undefined ->
+            %% maybe it's a dryrun connector
+            ets:lookup_element(?CACHE, ID, #connector.cb, undefined);
```

Comment:
Nit:

```suggestion
            ets:lookup_element(?CACHE, ID, #connector.cb, not_found);
```

Reply:
this is intended to be `undefined` to align with the type spec and other clauses

Reply:
that field seems to be just `term()` :thinking: 

it's just a minor nit, since everywhere else in this module we return `not_found` on such occasions

Reply:
sorry, i meant default value.
but i will push a new commit to use `no_state` instead.